docs/topics/db/optimization.txt

```diff
@@ -411,4 +416,105 @@ objects to reduce the number of SQL queries. For example::
     my_band.members.add(me)
     my_band.members.add(my_friend)
 
-...where ``Bands`` and ``Artists`` have a many-to-many relationship.
+...where ``Band`` and ``Artist`` have a many-to-many relationship.
+
+If you have a list of objects, use the ``*``-operator when calling
+:meth:`~django.db.models.fields.related.RelatedManager.add`::
+
+    my_band.members.add(*new_members)
+
+Many source objects, many target objects
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When inserting pairs of objects into :class:`ManyToManyFields
+<django.db.models.ManyToManyField>`, use 
+:meth:`~django.db.models.query.QuerySet.bulk_create()` with multiple
+:attr:`through <django.db.models.ManyToManyField.through>` model instances 
+to reduce the number of SQL queries. For example::
+
+    PizzaToppingRelation = Pizza.toppings.through
+    PizzaToppingRelation.objects.bulk_create([
+        PizzaToppingRelation(pizza=my_pizza, topping=pepperoni),
+        PizzaToppingRelation(pizza=your_pizza, topping=pepperoni),
+        PizzaToppingRelation(pizza=your_pizza, topping=mushroom),
+    ], ignore_conflicts=True)
+
+...is preferable to::
+
+    my_pizza.toppings.add(pepperoni)
+    your_pizza.toppings.add(pepperoni, mushroom)
+
+...where ``Pizza`` and ``Topping`` have a many-to-many relationship.
+
+If you have a list of object pairs, convert them to
+:attr:`through <django.db.models.ManyToManyField.through>`
+model instances first::
+
+    PizzaToppingRelation = Pizza.toppings.through
+    PizzaToppingRelation.objects.bulk_create([
+        PizzaToppingRelation(pizza=pizza, topping=topping)
+        for (pizza, topping) in
+        pizza_topping_tuples
+    ], ignore_conflicts=True)
```

Comment:
I don't think that we need an example how to convert list of objects into list to `through` instances. It seems unnecessary in the advanced topic about optimization.

Reply:
I have removed it, and its paired rather-involved Q() query construction in the bulk-remove section.

However I might highlight: **Now there is now neither documentation for how to explicitly add/remove relations based on a list of object pairs, nor is there a method that provides that functionality.**

The main thrust of the original contribution is no longer present :( , although I do think the docs have been improved in general.